x-pack/plugins/index_lifecycle_management/public/application/sections/edit_policy/components/phases/shared_fields/data_tier_allocation_field/components/node_allocation.tsx

```diff
@@ -60,6 +66,20 @@ export const NodeAllocation: FunctionComponent<SharedProps> = ({ phase, nodes, i
 
   nodeOptions.sort((a, b) => a.value.localeCompare(b.value));
 
+  let nodeAllocationOptions = [];
+
+  // On Cloud, allocating to data tiers and allocating to data nodes is mutually exclusive. So we
+  // only let users select this option if they're using data nodes. Otherwise we remove it.r
+  //
+  // On prem, users should have the freedom to choose this option, even if they're using node roles.
+  // So we always give them this option.
+  if (!isCloudEnabled || isUsingDeprecatedDataRoleConfig) {
```

Comment:
This is also a refactoring to make the condition easier to reason about locally. No behavior change.